File: Minimal/src/Brontide.Minimal.Host/ProviderRestartEnforcement.fs
namespace Brontide.Minimal.Host

open System

type ProviderRestartEnforcementResult =
    { Code: string
      RefusedBy: string
      Decision: ProviderRestartDecision
      Activation: ProviderServingActivation option
      ProviderStarted: bool
      LifecycleReconstructed: bool
      LogicalGenerationPreserved: bool }

/// Reconstructs a stopped provider connection from the activation's retained verified recipe.
[<RequireQualifiedAccess>]
module ProviderRestartEnforcement =
    let private result code refusedBy decision activation providerStarted lifecycleReconstructed logicalGenerationPreserved =
        { Code = code
          RefusedBy = refusedBy
          Decision = decision
          Activation = activation
          ProviderStarted = providerStarted
          LifecycleReconstructed = lifecycleReconstructed
          LogicalGenerationPreserved = logicalGenerationPreserved }

    let private runCore environment
        (policy: ProviderRestartPolicy)
        (registry: DurableProviderPublisherTrustPolicyRegistry)
        (store: ContentAddressedProviderStore)
        (activation: ProviderServingActivation)
        cause
        currentCyclePolicyIdentity
        now
        attemptCount
        lastAttempt = task {
        if isNull (box policy) then nullArg (nameof policy)
        if isNull (box registry) then nullArg (nameof registry)
        if isNull (box store) then nullArg (nameof store)
        if isNull (box activation) then nullArg (nameof activation)

        let decision =
            policy.Evaluate(
                registry, activation, cause, currentCyclePolicyIdentity,
                now, attemptCount, lastAttempt)

        if not decision.MayRestart then
            return result decision.Code decision.RefusedBy decision None false false false
        else
            let claim = activation.BeginRestart()
            if claim <> "restart-claimed" then
                return result claim "claim" decision None false false false
            else
                let mutable completed = false
                let mutable launchedProvider: StagedProviderProcess option = None
                try
                    match registry.Current, decision.PolicyIdentity with
                    | Some current, Some identity when current.Policy.Identity = identity ->
                        match activation.DistributionChain.Provider, activation.BindingLifecycle with
                        | None, _ | _, None ->
                            return result "provider-restart-activation-unavailable" "state" decision None false false false
                        | Some priorProvider, Some priorLifecycle ->
                            let staged = priorProvider.StagedArtifacts
                            match
                                if Map.isEmpty environment then store.Activate(staged, staged.Arguments)
                                else store.ActivateWithEnvironment(staged, staged.Arguments, environment)
                            with
                            | StagedProviderActivation.Refused failure ->
                                return result failure.Code "cbi31" decision None false false false
                            | StagedProviderActivation.Launched provider ->
                                launchedProvider <- Some provider
                                let chain =
                                    ProviderDistributionChain.restarted
                                        activation.DistributionChain provider identity
                                let! lifecycle =
                                    ComponentBindingLifecycle.restart
                                        priorLifecycle
                                        activation.RetainedResolution
                                        activation.RetainedSelection
                                        activation.RetainedRequest
                                        provider.Conversation
                                match lifecycle.Failure, lifecycle.Runtime, lifecycle.Member with
                                | Some failure, _, _ ->
                                    return result failure.Code "cbi2" decision None true false false
                                | None, Some runtime, Some memberValue
                                    when runtime.Kind = Brontide.Minimal.Experimental.ComponentManagement.ActivationRuntimeOutcomeKind.Active
                                         && memberValue.IsReleased ->
                                    let successor = activation.Restarted(chain, lifecycle)
                                    completed <- true
                                    return result "provider-restart-completed" "none" decision (Some successor) true true
                                        (lifecycle.Runtime = priorLifecycle.Runtime)
                                | _ ->
                                    return result "restart-lifecycle-incomplete" "cbi2" decision None true false false
                    | _ ->
                        return result "provider-restart-current-proof-required" "current-cycle" decision None false false false
                finally
                    if not completed then
                        launchedProvider |> Option.iter (fun provider -> provider.Dispose())
                    activation.FinishRestart completed
    }

    let run policy registry store activation cause currentCyclePolicyIdentity now attemptCount lastAttempt =
        runCore Map.empty policy registry store activation cause currentCyclePolicyIdentity now attemptCount lastAttempt

    let internal runWithEffectEnvironment environment policy registry store activation cause currentCyclePolicyIdentity now attemptCount lastAttempt =
        runCore environment policy registry store activation cause currentCyclePolicyIdentity now attemptCount lastAttempt

File: Reference/src/Brontide.Reference.Studio/ProviderRestartEnforcement.cs
namespace Brontide.Reference.Studio;

public sealed record ProviderRestartEnforcementResult(
    string Code,
    string RefusedBy,
    ProviderRestartDecision Decision,
    ProviderServingActivation? Activation,
    bool ProviderStarted,
    bool LifecycleReconstructed,
    bool LogicalGenerationPreserved);

/// <summary>
/// Reconstructs a stopped provider connection from the activation's retained verified recipe.
/// </summary>
public static class ProviderRestartEnforcement
{
    public static async ValueTask<ProviderRestartEnforcementResult> RunAsync(
        ProviderRestartPolicy policy,
        DurableProviderPublisherTrustPolicyRegistry registry,
        ContentAddressedProviderStore store,
        ProviderServingActivation activation,
        ProviderRestartCause cause,
        ProviderPublisherTrustPolicyId currentCyclePolicyIdentity,
        DateTimeOffset now,
        int attemptCount,
        DateTimeOffset? lastAttempt)
    {
        ArgumentNullException.ThrowIfNull(policy);
        ArgumentNullException.ThrowIfNull(registry);
        ArgumentNullException.ThrowIfNull(store);
        ArgumentNullException.ThrowIfNull(activation);

        var decision = policy.Evaluate(
            registry, activation, cause, currentCyclePolicyIdentity, now, attemptCount, lastAttempt);
        if (!decision.MayRestart)
            return Result(decision.Code, decision.RefusedBy, decision);

        var claim = activation.BeginRestart();
        if (claim != "restart-claimed")
            return Result(claim, "claim", decision);

        var completed = false;
        StagedProviderProcess? launchedProvider = null;
        try
        {
            var current = registry.Current;
            if (current is null || decision.PolicyIdentity is not { } policyIdentity
                || current.Policy.Identity != policyIdentity)
                return Result("provider-restart-current-proof-required", "current-cycle", decision);

            var priorChain = activation.Chain;
            if (priorChain.Provider is not { } priorProvider || activation.Lifecycle is not { } priorLifecycle)
                return Result("provider-restart-activation-unavailable", "state", decision);

            var staged = priorProvider.StagedArtifacts;
            var launched = store.Activate(staged, staged.Arguments);
            if (launched.Owner is not { } provider)
                return Result(launched.Failure!.Code, "cbi31", decision);
            launchedProvider = provider;

            var chain = ProviderDistributionChainResult.Restarted(priorChain, provider, policyIdentity);
            var lifecycle = await ComponentBindingLifecycle.RestartAsync(
                priorLifecycle,
                activation.Resolution,
                activation.Selection,
                activation.Request,
                provider.Conversation).ConfigureAwait(false);
            if (!lifecycle.IsActive)
            {
                if (lifecycle.Member is { } member)
                    await member.DisposeAsync().ConfigureAwait(false);
                return Result(
                    lifecycle.Failure?.Code ?? "restart-lifecycle-incomplete",
                    "cbi2", decision, providerStarted: true);
            }

            var successor = new ProviderServingActivation(
                chain, lifecycle, activation.Occurrence,
                activation.Resolution, activation.Selection, activation.Request);
            completed = true;
            return Result(
                "provider-restart-completed", "none", decision,
                successor, providerStarted: true, lifecycleReconstructed: true,
                logicalGenerationPreserved: lifecycle.Runtime == priorLifecycle.Runtime);
        }
        finally
        {
            if (!completed && launchedProvider is not null)
                await launchedProvider.DisposeAsync().ConfigureAwait(false);
            activation.FinishRestart(completed);
        }
    }

    private static ProviderRestartEnforcementResult Result(
        string code,
        string refusedBy,
        ProviderRestartDecision decision,
        ProviderServingActivation? activation = null,
        bool providerStarted = false,
        bool lifecycleReconstructed = false,
        bool logicalGenerationPreserved = false) =>
        new(code, refusedBy, decision, activation, providerStarted,
            lifecycleReconstructed, logicalGenerationPreserved);
}
